tools/codegen/gen.py

```diff
@@ -104,6 +104,38 @@ def cpp_string(s: str) -> str:
 # to be generated.  This pattern makes it convenient to use map, concatMap
 # and similar functional combinators.
 
+def static_dispatch_extra_headers(backends: Optional[List[DispatchKey]]) -> str:
+    if not backends:
+        return ''
+    return '\n'.join(f'#include <ATen/{key}Functions.h>' for key in backends) + """
+#include <ATen/DefaultBackendFunctions.h>
+#include <ATen/MathFunctions.h>
+"""
+
+def static_dispatch(
+    f: NativeFunction, cpp_sig: CppSignature,
+    *, method: bool, backends: Optional[List[DispatchKey]]
+) -> Optional[str]:
+    if not backends or f.manual_kernel_registration:
+        return None
+
+    assert len(backends) <= 1, "Multi-backend static dispatch is not yet supported!"
+    target_backend = backends[0]
+    target_sig = CppSignatureGroup.from_native_function(f, method=False, fallback_binding=False).signature
+    name = target_sig.name()
+    exprs = translate(cpp_sig.arguments(), target_sig.arguments(), method=method)
+    exprs_str = ', '.join(a.expr for a in exprs)
+
+    for dispatch_key in (target_backend, DispatchKey.DefaultBackend, DispatchKey.Math):
+        # TODO: for ops with structured_delegate it should check the dispatch table of
+        # the out variant instead. For now, these structured ops all have CPU/CUDA kernels
+        # so we always dispatch to the target_backend, but this could be wrong when we
+        # migrate math/default_backend ops to use structured delegate.
```

Comment:
Per our chat about emitting a warning in the TODO situation - if it's easy I'd throw it in (it seems like just an extra check based on the info you already have available here, right?)

Reply:
I thought about it again - I'm a little concerned it will generate a lot warnings when we start migrating more CPU ops to use structure kernels. The TODO situation should not be harmful as long as it compiles, i.e. when the structured-delegate actually has the CPU kernel it should be correct to call at::cpu::delegate_op(). If it doesn't have CPU kernel, then the build will fail anyway - we cannot print warning for the failure case unless we actually query the structured_delegate's dispatch table (basically what TODO needs to do).

I moved this TODO out of the for loop to make it more clear.